custom programmer broken in v7.3

If I checkout v7.2 and build it on MacOSX, I can have the following in `~/.avrduderc`:
```
programmer
  id    = "ft232r-cts";
  desc  = "FT232R Synchronous BitBang, CTS for RESET";
  type  = "ftdi_syncbb";
  connection_type = usb;
  sck   = 0;  # TxD
  #sdi   = 1;  # RxD
  #sdo   = 2;  # RTS
  miso  = 1;  # RxD
  mosi  = 2;  # RTS
  reset = 3;  # CTS
  #reset = 4;  # DTR
;

#      PIN NO.     DATA LINE
#       0           TXD
#       1           RXD
#       2           RTS
#       3           CTS
#       4           DTR
#       5           DSR
#       6           DCD
#       7           RI
```
and 7.2 sees it fine:
```
$ git checkout v7.2
[...]
$ sh build.sh -j 4 && ./build_darwin/src/avrdude -v -p attiny85 -c ft232r-cts
[...]
avrdude: Version 7.2-20230719 (cb9e7e49)
[...]
         System wide configuration file is /Users/jmg/github/avrdude/build_darwin/src/avrdude.conf
         User configuration file is /Users/jmg/.avrduderc
avrdude yywarning() warning: miso is deprecated, will be removed in v8.0, use sdi [/Users/jmg/.avrduderc:9]
avrdude yywarning() warning: mosi is deprecated, will be removed in v8.0, use sdo [/Users/jmg/.avrduderc:10]

         Using Port                    : usb
         Using Programmer              : ft232r-cts
[...]
avrdude ft245r_program_enable() error: device is not responding to program enable; check connection
avrdude main() error: initialization failed, rc=-1
        - double check the connections and try again
```

But if I do the same for 7.3, the programmer is not found:
```
$ git checkout v7.3
Previous HEAD position was cb9e7e49 Prepare for version 7.2
HEAD is now at e599214c Update version info for 7.3
$ sh build.sh && ./build_darwin/src/avrdude -v -p attiny85 -c ft232r-cts
[...]
avrdude: Version 7.3
[...]
         System wide configuration file is /Users/jmg/github/avrdude/build_darwin/src/avrdude.conf
         User configuration file is /Users/jmg/.avrduderc
avrdude yywarning() warning: miso is deprecated, will be removed in v8.0, use sdi [/Users/jmg/.avrduderc:9]
avrdude yywarning() warning: mosi is deprecated, will be removed in v8.0, use sdo [/Users/jmg/.avrduderc:10]

avrdude programmer_not_found() error: cannot find programmer id ft232r-cts
use -c? to see all possible programmers
```

and the programmer ft232r-cts is not seen in `-c?` for 7.3, but it is for 7.2.

After a bit of digging:
```
avrdude: Version 7.2-20230827 (b8c47f39)
```
"works" in that the programmer is seen, but it fails with:
```
avrdude main() error: programmer ft232r-cts cannot program part ATtiny85 as they
        lack a common programming mode; use -F to override this check
```

It identifies commit 91ffab5fc6f22 as the bad one, but 91ffab5fc6f22 fails to compile:
```
CMake Error at src/CMakeLists.txt:140 (add_library):
  Cannot find source file:

    serialadapter.c
```

so the breakage was introduced in commits 91ffab5fc6f2 and aaddbbf18731.

--- a/src/config_gram.y
+++ b/src/config_gram.y
@@ -275,6 +275,9 @@ prog_def :
         yyerror("programmer type not specified");
         YYABORT;
       }
+      if(!current_prog->prog_modes && cx->lex_kw_is_programmer)
+        yywarning("programmer %s fails to specify prog_modes = PM_...;",
+          (char *) ldata(lfirst(current_prog->id)));
       for(LNODEID ln = lfirst(current_prog->id); ln; ln = lnext(ln)) {
         char *id = ldata(ln);
         if((existing_prog = locate_programmer(programmers, id))) {

--- a/src/lexer.l
+++ b/src/lexer.l
@@ -226,7 +226,8 @@ part             { yylval=NULL; ccap(); current_strct = COMP_AVRPART; return K_P
 pgm_enable       { yylval=new_token(K_PGM_ENABLE); ccap(); return K_PGM_ENABLE; }
 pgmled           { yylval=NULL; ccap(); return K_PGMLED; }
 pp_controlstack  { yylval=NULL; ccap(); return K_PP_CONTROLSTACK; }
-(programmer|serialadapter) { yylval=NULL; ccap(); current_strct = COMP_PROGRAMMER; return K_PROGRAMMER; }
+(programmer|serialadapter) { yylval=NULL; ccap(); current_strct = COMP_PROGRAMMER;
+                   cx->lex_kw_is_programmer = *yytext == 'p'; return K_PROGRAMMER; }
 rdyled           { yylval=NULL; ccap(); return K_RDYLED; }
 read             { yylval=new_token(K_READ); ccap(); return K_READ; }
 read_hi          { yylval=new_token(K_READ_HI); ccap(); return K_READ_HI; }

--- a/src/libavrdude.h
+++ b/src/libavrdude.h
@@ -1581,7 +1581,7 @@ typedef struct {
   uint16_t dfu_wIndex;          // A running number for USB messages
 
   // Static variable from config_gram.y
- int cfgy_pin_name;             // Temporary variable for grammar parsing
+  int cfgy_pin_name;            // Temporary variable for grammar parsing
 
   // Static variable from ppi.c
   unsigned char ppi_shadow[3];
@@ -1622,6 +1622,9 @@ typedef struct {
   char usb_buf[USBDEV_MAX_XFER_3];
   int usb_buflen, usb_bufptr; // @@@ Check whether usb_buflen needs initialising with -1
   int usb_interface;
+
+  // Variable connecting lexer.l and config_gram.y
+  int lex_kw_is_programmer;     // Was the K_PROGRAMMER keyword "programmer"?
 } libavrdude_context;
 
 extern libavrdude_context *cx;

